File: Lab2/Main.java
import java.util.ArrayList;
import java.util.List;

// 1. HERENCIA: Clase Base
class Persona {
    private String nombre;
    private int edad;

    public Persona() {}

    public Persona(String nombre, int edad) {
        this.nombre = nombre;
        this.edad = edad;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public int getEdad() {
        return edad;
    }

    public void setEdad(int edad) {
        this.edad = edad;
    }

    @Override
    public String toString() {
        return "Persona [nombre=" + nombre + ", edad=" + edad + "]";
    }
}

// 1. HERENCIA: Clases Hijas
class Profesor extends Persona {
    private String especialidad;

    public Profesor() {
        super();
    }

    public Profesor(String nombre, int edad, String especialidad) {
        super(nombre, edad);
        this.especialidad = especialidad;
    }

    public String getEspecialidad() {
        return especialidad;
    }

    public void setEspecialidad(String especialidad) {
        this.especialidad = especialidad;
    }

    @Override
    public String toString() {
        return "Profesor [" + super.toString() + ", especialidad=" + especialidad + "]";
    }
}

class Estudiante extends Persona {
    private String codigoEstudiante;

    public Estudiante() {
        super();
    }

    public Estudiante(String nombre, int edad, String codigoEstudiante) {
        super(nombre, edad);
        this.codigoEstudiante = codigoEstudiante;
    }

    public String getCodigoEstudiante() {
        return codigoEstudiante;
    }

    public void setCodigoEstudiante(String codigoEstudiante) {
        this.codigoEstudiante = codigoEstudiante;
    }

    @Override
    public String toString() {
        return "Estudiante [" + super.toString() + ", codigo=" + codigoEstudiante + "]";
    }
}

// 2. COMPOSICIÓN: Horario es parte vital de Curso
class Horario {
    private String dias;
    private String horas;

    public Horario() {}

    public Horario(String dias, String horas) {
        this.dias = dias;
        this.horas = horas;
    }

    public String getDias() {
        return dias;
    }

    public void setDias(String dias) {
        this.dias = dias;
    }

    public String getHoras() {
        return horas;
    }

    public void setHoras(String horas) {
        this.horas = horas;
    }

    @Override
    public String toString() {
        return "Horario [dias=" + dias + ", horas=" + horas + "]";
    }
}

class Curso {
    private String nombre;
    // Composición: el Horario nace y está íntimamente ligado al Curso
    private Horario horario;

    public Curso(String nombre, String dias, String horas) {
        this.nombre = nombre;
        // Se instancia directamente para evidenciar la composición fuerte
        this.horario = new Horario(dias, horas);
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public Horario getHorario() {
        return horario;
    }

    public void setHorario(Horario horario) {
        this.horario = horario;
    }

    @Override
    public String toString() {
        return "Curso [nombre=" + nombre + ", " + horario.toString() + "]";
    }
}

// 3. AGREGACIÓN: Universidad contiene Cursos independientes
class Universidad {
    private String nombre;
    private List<Curso> listaCursos;

    public Universidad() {
        this.listaCursos = new ArrayList<>();
    }

    public Universidad(String nombre) {
        this.nombre = nombre;
        this.listaCursos = new ArrayList<>();
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public List<Curso> getListaCursos() {
        return listaCursos;
    }

    // Agregación: Recibe cursos ya existentes exteriormente
    public void agregarCurso(Curso curso) {
        this.listaCursos.add(curso);
    }

    @Override
    public String toString() {
        return "Universidad [nombre=" + nombre + ", cantidadCursos=" + listaCursos.size() + "]";
    }
}

// 4. DEPENDENCIA: Reporte usa temporalmente a Estudiante
class Reporte {
    public Reporte() {}

    // Dependencia: Estudiante pasa como parámetro temporal para la operación
    public void generarReporteEstudiante(Estudiante estudiante) {
        System.out.println("==========================================");
        System.out.println("       REPORTE ACADÉMICO DE ESTUDIANTE    ");
        System.out.println("==========================================");
        System.out.println("Nombre: " + estudiante.getNombre());
        System.out.println("Edad:   " + estudiante.getEdad() + " años");
        System.out.println("Código: " + estudiante.getCodigoEstudiante());
        System.out.println("Estado: ACTIVO REGULAR");
        System.out.println("==========================================\n");
    }

    @Override
    public String toString() {
        return "Reporte [modulo=Generador de reportes temporales]";
    }
}

// PROGRAMA PRINCIPAL (Main)
public class Main {
    public static void main(String[] args) {
        System.out.println("        SISTEMA DE GESTIÓN UNIVERSITARIA          ");
        System.out.println("\n");

        // 1. Creación de 2 profesores (Herencia)
        Profesor prof1 = new Profesor("Dr. Carlos Ruiz", 48, "Sistemas Distribuidos");
        Profesor prof2 = new Profesor("MSc. Ana Paredes", 39, "Arquitectura de Software");

        System.out.println("--- Profesores Registrados ---");
        System.out.println(prof1);
        System.out.println(prof2);
        System.out.println();

        // 2. Creación de 3 estudiantes (Herencia)
        Estudiante est1 = new Estudiante("Juan Perez", 20, "20220101");
        Estudiante est2 = new Estudiante("Maria Lopez", 22, "20210452");
        Estudiante est3 = new Estudiante("Luis Quispe", 21, "20221189");

        System.out.println("--- Estudiantes Registrados ---");
        System.out.println(est1);
        System.out.println(est2);
        System.out.println(est3);
        System.out.println();

        // 3. Creación de 2 cursos (Composición con Horario)
        Curso curso1 = new Curso("Tecnologia de Objetos", "Lunes y Miercoles", "08:00 - 10:00");
        Curso curso2 = new Curso("Bases de Datos Avanzadas", "Martes y Jueves", "14:00 - 16:00");

        System.out.println("--- Cursos Creados (con Horario por Composición) ---");
        System.out.println(curso1);
        System.out.println(curso2);
        System.out.println();

        // 4. Agregar cursos a la Universidad (Agregación)
        Universidad uni = new Universidad("Universidad Nacional Tecnológica");
        uni.agregarCurso(curso1);
        uni.agregarCurso(curso2);

        System.out.println("--- Universidad y Cursos Agregados ---");
        System.out.println(uni);
        for (Curso c : uni.getListaCursos()) {
            System.out.println(" -> " + c);
        }
        System.out.println();

        // 5. Generar reporte de un estudiante (Dependencia)
        Reporte reporte = new Reporte();
        reporte.generarReporteEstudiante(est1);
    }
}

File: Lab2/Main.py
import typing
from typing import List

# --- 1. HERENCIA ---
# Clase Base
class Persona:
    def __init__(self, nombre: str = "", edad: int = 0) -> None:
        # Usamos un guion bajo para simular el 'private' de Java
        self._nombre: str = nombre
        self._edad: int = edad

    def get_nombre(self) -> str:
        return self._nombre

    def set_nombre(self, nombre: str) -> None:
        self._nombre = nombre

    def get_edad(self) -> int:
        return self._edad

    def set_edad(self, edad: int) -> None:
        self._edad = edad

    # Equivalente al toString()
    def __str__(self) -> str:
        return f"Persona [nombre={self._nombre}, edad={self._edad}]"

# Clases Hijas que heredan de Persona
class Profesor(Persona):
    def __init__(self, nombre: str = "", edad: int = 0, especialidad: str = "") -> None:
        # Llamamos al constructor de la clase padre
        super().__init__(nombre, edad)
        self._especialidad: str = especialidad

    def get_especialidad(self) -> str:
        return self._especialidad

    def set_especialidad(self, especialidad: str) -> None:
        self._especialidad = especialidad

    def __str__(self) -> str:
        return f"Profesor [{super().__str__()}, especialidad={self._especialidad}]"

class Estudiante(Persona):
    def __init__(self, nombre: str = "", edad: int = 0, codigo_estudiante: str = "") -> None:
        super().__init__(nombre, edad)
        self._codigo_estudiante: str = codigo_estudiante

    def get_codigo_estudiante(self) -> str:
        return self._codigo_estudiante

    def set_codigo_estudiante(self, codigo_estudiante: str) -> None:
        self._codigo_estudiante = codigo_estudiante

    def __str__(self) -> str:
        return f"Estudiante [{super().__str__()}, codigo={self._codigo_estudiante}]"

# --- 2. COMPOSICIÓN ---
# El Horario no tiene sentido sin un Curso
class Horario:
    def __init__(self, dias: str = "", horas: str = "") -> None:
        self._dias: str = dias
        self._horas: str = horas

    def get_dias(self) -> str:
        return self._dias

    def set_dias(self, dias: str) -> None:
        self._dias = dias

    def get_horas(self) -> str:
        return self._horas

    def set_horas(self, horas: str) -> None:
        self._horas = horas

    def __str__(self) -> str:
        return f"Horario [dias={self._dias}, horas={self._horas}]"

class Curso:
    def __init__(self, nombre: str, dias: str, horas: str) -> None:
        self._nombre: str = nombre
        # Aquí se ve la composición: instanciamos el Horario dentro del Curso
        self._horario: Horario = Horario(dias, horas)

    def get_nombre(self) -> str:
        return self._nombre

    def set_nombre(self, nombre: str) -> None:
        self._nombre = nombre

    def get_horario(self) -> Horario:
        return self._horario

    def set_horario(self, horario: Horario) -> None:
        self._horario = horario

    def __str__(self) -> str:
        return f"Curso [nombre={self._nombre}, {self._horario.__str__()}]"

# --- 3. AGREGACIÓN ---
# La Universidad agrupa Cursos, pero los cursos existen por su cuenta
class Universidad:
    def __init__(self, nombre: str = "") -> None:
        self._nombre: str = nombre
        self._lista_cursos: List[Curso] = []

    def get_nombre(self) -> str:
        return self._nombre

    def set_nombre(self, nombre: str) -> None:
        self._nombre = nombre

    def get_lista_cursos(self) -> List[Curso]:
        return self._lista_cursos

    # Recibimos el curso desde afuera y lo agregamos a la lista
    def agregar_curso(self, curso: Curso) -> None:
        self._lista_cursos.append(curso)

    def __str__(self) -> str:
        return f"Universidad [nombre={self._nombre}, cantidadCursos={len(self._lista_cursos)}]"

# --- 4. DEPENDENCIA ---
# El Reporte solo usa al Estudiante un rato para imprimir sus datos
class Reporte:
    def __init__(self) -> None:
        pass

    # Pasamos al estudiante como parámetro, lo usamos y ya
    def generar_reporte_estudiante(self, estudiante: Estudiante) -> None:
        print("==========================================")
        print("       REPORTE ACADÉMICO DE ESTUDIANTE    ")
        print("==========================================")
        print(f"Nombre: {estudiante.get_nombre()}")
        print(f"Edad:   {estudiante.get_edad()} años")
        print(f"Código: {estudiante.get_codigo_estudiante()}")
        print("Estado: ACTIVO REGULAR")
        print("==========================================\n")

    def __str__(self) -> str:
        return "Reporte [modulo=Generador de reportes temporales]"

# --- PROGRAMA PRINCIPAL ---
def main() -> None:
    print("        SISTEMA DE GESTIÓN UNIVERSITARIA          ")
    print("\n")

    # 1. Herencia: Creando profesores y estudiantes
    prof1: Profesor = Profesor("Dr. Carlos Ruiz", 48, "Sistemas Distribuidos")
    prof2: Profesor = Profesor("MSc. Ana Paredes", 39, "Arquitectura de Software")

    print("--- Profesores Registrados ---")
    print(prof1)
    print(prof2)
    print()

    est1: Estudiante = Estudiante("Juan Perez", 20, "20220101")
    est2: Estudiante = Estudiante("Maria Lopez", 22, "20210452")
    est3: Estudiante = Estudiante("Luis Quispe", 21, "20221189")

    print("--- Estudiantes Registrados ---")
    print(est1)
    print(est2)
    print(est3)
    print()

    # 2. Composición: Al crear el curso, internamente nace su horario
    curso1: Curso = Curso("Tecnologia de Objetos", "Lunes y Miercoles", "08:00 - 10:00")
    curso2: Curso = Curso("Bases de Datos Avanzadas", "Martes y Jueves", "14:00 - 16:00")

    print("--- Cursos Creados (con Horario por Composición) ---")
    print(curso1)
    print(curso2)
    print()

    # 3. Agregación: Metemos los cursos ya creados a la universidad
    uni: Universidad = Universidad("Universidad Nacional Tecnológica")
    uni.agregar_curso(curso1)
    uni.agregar_curso(curso2)

    print("--- Universidad y Cursos Agregados ---")
    print(uni)
    for c in uni.get_lista_cursos():
        print(f" -> {c}")
    print()

    # 4. Dependencia: Generamos el reporte de un alumno
    reporte: Reporte = Reporte()
    reporte.generar_reporte_estudiante(est1)

if __name__ == "__main__":
    main()
